src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -156,7 +156,123 @@ bool DiscoveryDataBase::process_data_queue()
+        std::vector<fastrtps::rtps::GUID_t> readers = readers_by_topic_[*topic_it];
+
+        for (fastrtps::rtps::GUID_t writer: writers)
+        // Iterate over writers in the topic:
+        {
+            // Iterate over readers in the topic:
+            for (fastrtps::rtps::GUID_t reader : readers)
+            {
+                // Find participants with writer info and participant with reader info in participants_
+                parts_reader_it = participants_.find(reader.guidPrefix);
+                parts_writer_it = participants_.find(writer.guidPrefix);
+                // Find reader info in readers_
+                readers_it = readers_.find(reader);
+                // Find writer info in writers_
+                writers_it = writers_.find(writer);
+
+                // Check in `participants_` whether the client with the reader has acknowledge the PDP of the client
+                // with the writer.
+                if (parts_reader_it != participants_.end() && parts_reader_it->second.is_matched(writer.guidPrefix))
+                {
+                    // Check the status of the writer in `readers_[reader]::relevant_participants_builtin_ack_status`.
+                    if (readers_it != readers_.end() && !readers_it->second.is_matched(writer.guidPrefix))
+                    {
+                        // If the status is 0, add DATA(w) to a `edp_publications_to_send_` (if it's not there).
+                        if (std::find(
+                            edp_publications_to_send_.begin(),
+                            edp_publications_to_send_.end(),
+                            writers_it->second.change()) == edp_publications_to_send_.end())
+                        {
+                            edp_publications_to_send_.push_back(writers_it->second.change());
+                        }
+                    }
+                }
+                else
+                {
+                    // Add DATA(p) of the client with the writer to `pdp_to_send_` (if it's not there).
+                    if (std::find(
+                        pdp_to_send_.begin(),
+                        pdp_to_send_.end(),
+                        parts_writer_it->second.change()) == pdp_to_send_.end())
+                    {
+                        pdp_to_send_.push_back(parts_writer_it->second.change());
+                    }
+                    // Set topic as not-clearable.
+                    is_clearable = false;
+                }
+
+                // Check in `participants_` whether the client with the writer has acknowledge the PDP of the client
+                // with the reader.
+                if (parts_writer_it != participants_.end() && parts_writer_it->second.is_matched(reader.guidPrefix))
+                {
+                    // Check the status of the reader in `writers_[writer]::relevant_participants_builtin_ack_status`.
+                    auto writers_it = writers_.find(writer);
```

Comment:
Erase this line. These declarations has been moved to the beginning of the function.

Reply:
:+1: